### ISSUE TITLE
Token became writable after getting error in readable on RST answer.

### ISSUE THREAD

**Issue by hoxnox** [carllerche/mio#179]
In the folowing example:

    extern crate mio;
    use std::net;
    
    struct Hnd {
        stream: Option<mio::tcp::TcpStream>
    }
    
    impl Hnd {
        fn new() -> Hnd {
            Hnd {
                stream: None
            }
        }
    }
    
    impl mio::Handler for Hnd {
        type Timeout = u32;
        type Message = ();
    
        fn timeout(&mut self, event_loop: &mut mio::EventLoop<Self>, timeout: u32) {
            println!("Timeout.");
            if timeout == 0 {
                event_loop.timeout_ms(0, 1000).unwrap();
            }
            if !self.stream.is_some() {
                let google = net::Ipv4Addr::new(49, 94, 146, 13);
                //let google = net::Ipv4Addr::new(173, 194, 122, 224);
                let stream = mio::tcp::TcpStream::connect(&net::SocketAddr::V4(net::SocketAddrV4::new(google, 80))).unwrap();
                event_loop.register_opt(&stream, mio::Token(0), mio::Interest::writable(), mio::PollOpt::edge()).unwrap();
                self.stream = Some(stream);
            }
        }
    
        fn writable(&mut self, _: &mut mio::EventLoop<Self>, token: mio::Token) {
            println!("Writable token {}.", token.as_usize());
        }
        fn readable(&mut self, _: &mut mio::EventLoop<Self>, token: mio::Token, rh: mio::ReadHint)
        {
            println!("Readable token {} with read hint {:?}, {:?}", token.as_usize(), rh.is_data(), rh.is_error());
        }
    }
    
    fn main() {
        let mut h = Hnd::new();
        let mut event_loop: mio::EventLoop<Hnd> = mio::EventLoop::new().unwrap();
        event_loop.timeout_ms(0, 1000).unwrap();
        event_loop.run(&mut h).unwrap();
    }
    
You'll have:

    ~/devel/miotest [master] ⚡ ./target/debug/miotest
    Timeout.
    Readable token 0 with read hint false, true
    Writable token 0.
    Timeout.

So after getting error in readable stream is writable, which I think is wrong.

**Comment by hoxnox** [comment]
It happens even if I deregister token in readable callback:

        fn readable(&mut self, event_loop: &mut mio::EventLoop<Self>, token: mio::Token, rh: mio::ReadHint)
        {
            println!("Readable token {} with read hint {:?}, {:?}", token.as_usize(), rh.is_data(), rh.is_error());
            let stream = match self.stream {
                Some(ref s) => s,
                None => {
                    println!("Error! No stream found.");
                    return;
                }
            };
            match event_loop.deregister(stream) {
                Err(e) => {
                    println!("Error deregistering stream. {:?}", e);
                    return;
                }
                _ => (),
            };
        }

Result:

    ~/devel/miotest [master] ⚡ ./target/debug/miotest
    Timeout.
    Readable token 0 with read hint false, true
    Writable token 0.
    Timeout.

**Comment by rrichardson** [comment]
I assume you are testing this on Linux?  This is the standard behavior for epoll.  If you connect() with a non-blocking socket and register it with epoll.  Epoll will send back whatever events you registered along with POLLERR.  In your case, you registered POLLIN (readable) and POLLOUT (writable), then it also provides the error hint indicating that your connect failed.   If you think about it, this is the correct behavior, as how else is the kernel going to notify you that your connection failed? 

After polling, mio checks the readable event, then calls Handler::readable, then it checks the writable event then calls Handler::writable.  https://github.com/carllerche/mio/blob/master/src/event_loop.rs#L296-L306

Since since you registered for both events, you got both events.  

@carllerche :  This is the 2nd time in a week that this behavior has tripped somebody up.  I recommend one of two options (I'm leaning towards the first, but the 2nd could be slightly more efficient): 

1.  Pass the ReadHint to the writable event as well so people can check for error/hup. (and change the name to EventHint or something) 

2. Use some standard event callback chaining logic which proceeds to the next event iff a person says proceed.  Maybe a binary result from readable and writable would be good enough, if the handler returns true, you proceed to writable, if the handler returns false, you stop for that file descriptor.

**Comment by hoxnox** [comment]
@rrichardson Yes, I'm using Linux and situation not quite normal - on normal epoll loop you got epoll_event and manually checks EPOLLERR, EPOLLIN, or whatever. In most cases you reacts only once, than continues to another epoll_event, or event breaks the loop. In case of mio - on error you have no options - readable and writable will be called one after another. And there is no built in mechanisms to check was the error already handled, or not. I think user should have ability not to pass error further. More over - I think error shouldn't pass further if it was handled by writable or readable.

Maybe add another callback - error?

**Comment by carllerche** [comment]
Closing in favor of #184
